### DISCUSSION TITLE
[FEATURE REQ] Further windows support (Tray mode, Updater)

### DISCUSSION THREAD

**Discussion by austinwbest** [autobrr/autobrr#1661] [Ideas]
Currently brr works fine on windows, so this is nothing more than a request on how to run it better as a non service app. 

**Tray mode**
Nearly every windows app has the ability (if not by default) to run in the tray hidden from the task bar at the bottom. Currently if you run the exe it opens a CMD window that is perm open in the task bar. This is console mode and is also handy for troubleshooting, following logs, etc but as this is a daemon window there really is no need to have it open and makes way more sense to "minimize" to the tray (all the Starr apps for example).

This request is to simply add another exe (if you want to keep the console one) that is ran in the tray. This does not require any advanced installer, MSI, UAC, etc work to be done. Cap has done this with Unpackerr and Notifiarr and neither have a full on installer or UAC issues. It is simply an exe with a conf file that is run. When opening it, it runs in the tray with a dynamic context menu (dont think that is needed for this project) and nothing more.

I'll have captain chime in and maybe he can take a look at this and give some suggestions on how to implement it so it isn't so much work or learning curve, etc for others who dont typically work with windows or want to for that matter as he has been through the headaches already

**Update check**
If not a full on auto updater (Notifiarr does this), at least a way to check for updates and have them download (Unpackerr does this by looking at GH releases) to be installed by a user.

**Comment by davidnewhall** [comment]
While there are probably several ways to get a tray in Windows, and hide the ever-running command window, I would like to give you my perspective after going through all the tribulations figuring this out. I'll stick mostly to things I learned along the way and not get too prescriptive on how you tackle the problem. I'm always here if you have questions, and happy to help!

- When you add a tray icon, it can support Mac, Windows, and Linux. You will have to make a determination at the time how you want this app to run on all these platforms. The tray generally does not work on other unix OSes (like BSD).
- I made very deliberate decisions to support Mac with and without the tray running. My apps can be installed via homebrew (no tray) or as a signed Mac .app (which has a tray). In hindsight, I would not have done this and would have jumped right to a signed app, and I would have made homebrew compile the .app instead of of a solid binary.
- I made the decision to not support the tray on Linux at all. I would suggest the same to you, but it's a personal choice you need to figure out.
- When I first made the tray, I did not have a web UI yet, so I was trying to put a lot of stuff into the tray to control the app. I do not recommend this. The tray should be short and simple. I recommend three Manu items and no/few more: `Open UI`, `Check for Update`, and `Quit`. The reason for this is to just keep your Windows-only code small and simple. 
- Some folks think that adding `-ldflags -H=windowsgui` to their app's compile options is the best way to hide the console window on Windows. And they may be correct, but I've found that having the ability to re-open the console window is a life saver. I set my apps up to [close the window immediately on start](https://github.com/Notifiarr/notifiarr/blob/main/main.go#L26), and they [re-open it inside `recover()`](https://github.com/Notifiarr/notifiarr/blob/main/pkg/logs/logs.go#L168-L176) if there's a panic. Then the operator can see the panic message in the window that 'appears'. If you add `-H=windowsgui` you will never be able to open the console window. If you choose to do this, you should remove all code that `prints` to stdout since it will never be seen again.
- The tray MUST run in the main go routine. You will likely bash your head against as wall as you try to make it not do that. The tray handles most of the exit routines, so plan for your app to have to "deal with the tray" when exiting. My apps often have 1 exit method that the tray calls before quitting, but it can also be called by other exit calls within the app. Do not expect this to go quickly, it will take you some time to figure out how to handle exiting gracefully with the tray code hogging your main go routine.

**Comment by klepp0906** [comment]
+1 on the vote for a tray icon.  makes it a bit easier to toss it in your startup folder as opposed to creating a task, and makes access to the webui substantially cleaner/more convenient.   most importantly, this will require the exe to get the bunny icon!

**Comment by sam272** [comment]
+1 for an auto-updater like the *arrs have

**Comment by austinwbest** [comment]
As ze0s doesnt like us windows users lol, hopefully this'll help in the meantime. 

- Create a file (i have it in desktop/scripts/) and name it autobrr_updater.py (or whatever you want) and paste the contents below in it.
- Make sure you have python installed and the 3 packages listed at the top
- Change the `autobrrURL`, `autobrrApikey`, `autobrrInstallFolder` and `autobrrShortcut` to your setup
- Run it manually to make sure it works.. Start -> cmd and then browse to wherever you put the py file and execute it (cd Desktop/scripts/ and once in the folder run ./autobrr_updater.py)
- Create a task scheduler entry to run it however often you want to check (mine is daily)

```
# pip install requests
# pip install packaging
# pip install psutil

import os
import shutil
import tempfile
import requests
import tarfile
import psutil
import subprocess
from packaging import version

print('Update check starting...');

autobrrURL              = 'http://localhost:7474'
autobrrApikey           = ''
autobrrInstallFolder    = 'C:\\ProgramData\\autobrr'
autobrrShortcut         = 'C:\\Users\\nitsua\\AppData\\Roaming\\Microsoft\\Windows\\Start Menu\\Programs\\Startup\\autobrr.lnk';

print('Connecting to autobrr on ' + autobrrURL + ' using apikey ' + autobrrApikey + '...')
autobrrConfig           = requests.get(autobrrURL + '/api/config?apikey=' + autobrrApikey)
autobrrConfigResponse   = autobrrConfig.json()
currentVersion          = autobrrConfigResponse['version']
print('Current autobrr version: ' + currentVersion)

releases = requests.get('https://api.github.com/repos/autobrr/autobrr/releases')
json = releases.json()

for release in json:
    releaseVersion = release['tag_name'].replace('v', '')
    print('Newest version of autobrr found: ' + releaseVersion)

    if version.parse(releaseVersion) > version.parse(currentVersion):
        print('Newest version is greater than current version, upgrade needed...')

        asset = 'https://github.com/autobrr/autobrr/releases/download/' + release['tag_name'] + '/autobrr_' + releaseVersion + '_windows_x86_64.zip'
        print('Grabbing asset: ' + asset + '...');

        tmpFolder           = tempfile.gettempdir()
        autobrrTmpFolder    = os.path.join(tmpFolder, 'autobrr')
        
        if os.path.exists(autobrrTmpFolder):
            shutil.rmtree(tmpFolder + '\\autobrr')

        os.mkdir(autobrrTmpFolder)
        tmpFilename         = 'autobrr_' + releaseVersion + '_windows_x86_64.zip'
        tmpFile             = tmpFolder + '\\autobrr\\' + tmpFilename

        print('Temp folder: ' + tmpFolder)
        print('Temp file: ' + tmpFilename)
        print('Using tmpfile to save: ' + tmpFile + '...')

        grabAsset = requests.get(asset)
        with open(tmpFile, 'wb') as a:
            print('Saving asset to tmpfile (' + grabAsset.headers['Content-length'] + ' B)...')
            a.write(grabAsset.content)

        print('Save finished, extracting asset...')
        shutil.unpack_archive(tmpFile, tmpFolder + '\\autobrr')
        print('Extracting finished, killing autobrr process...')

        process = 'autobrr.exe'
        for proc in psutil.process_iter():
            if proc.name() == process:
                proc.kill()
        print('Process killed, moving update files...')

        shutil.move(tmpFolder + '\\autobrr\\autobrr.exe', autobrrInstallFolder + '\\autobrr.exe')
        shutil.move(tmpFolder + '\\autobrr\\autobrrctl.exe', autobrrInstallFolder + '\\autobrrctl.exe')
        print('Files moved, restarting autobrr...')
        
        os.startfile(autobrrShortcut)
        print('Autobrr restarted, cleaning up...')

        shutil.rmtree(tmpFolder + '\\autobrr')
        break

print('Update check complete.')
exit()
```

**Comment by bulletproof2k** [comment]
+1 👍
